File: client/uptunnel/protocol.py
"""Wire protocol v1 — see docs/PROTOCOL.md.

Deliberately dependency-free and written with no comprehensions over large buffers so the
same file can be lifted into MicroPython for an ESP32 client later.
"""

import json
import struct

PROTOCOL_VERSION = 1

HELLO = 0x01
HELLO_OK = 0x02
ERROR = 0x03
OPEN_TUNNEL = 0x10
TUNNEL_OK = 0x11
CLOSE_TUNNEL = 0x12
STREAM_OPEN = 0x20
STREAM_DATA = 0x21
STREAM_EOF = 0x22
STREAM_ACK = 0x23
STREAM_RESET = 0x24

STREAM_FRAME_FLOOR = 0x20

NAMES = {
    HELLO: "HELLO",
    HELLO_OK: "HELLO_OK",
    ERROR: "ERROR",
    OPEN_TUNNEL: "OPEN_TUNNEL",
    TUNNEL_OK: "TUNNEL_OK",
    CLOSE_TUNNEL: "CLOSE_TUNNEL",
    STREAM_OPEN: "STREAM_OPEN",
    STREAM_DATA: "STREAM_DATA",
    STREAM_EOF: "STREAM_EOF",
    STREAM_ACK: "STREAM_ACK",
    STREAM_RESET: "STREAM_RESET",
}


class ProtocolError(Exception):
    pass


def control(frame_type: int, body: dict) -> bytes:
    return bytes([frame_type]) + json.dumps(body).encode("utf-8")


def stream_json(frame_type: int, stream_id: int, body: dict) -> bytes:
    return (
        bytes([frame_type])
        + struct.pack(">I", stream_id)
        + json.dumps(body).encode("utf-8")
    )


def stream_data(stream_id: int, payload: bytes) -> bytes:
    return bytes([STREAM_DATA]) + struct.pack(">I", stream_id) + payload


def stream_only(frame_type: int, stream_id: int) -> bytes:
    return bytes([frame_type]) + struct.pack(">I", stream_id)


def stream_ack(stream_id: int, consumed: int) -> bytes:
    return bytes([STREAM_ACK]) + struct.pack(">II", stream_id, consumed)


def decode(buf: bytes):
    """Returns (frame_type, stream_id_or_None, body_or_payload).

    Control frames yield a parsed dict; stream frames yield the raw payload bytes.
    """
    if not buf:
        raise ProtocolError("empty frame")
    frame_type = buf[0]

    if frame_type < STREAM_FRAME_FLOOR:
        if len(buf) == 1:
            return frame_type, None, {}
        try:
            body = json.loads(buf[1:].decode("utf-8"))
        except (ValueError, UnicodeDecodeError) as exc:
            raise ProtocolError("malformed JSON in frame 0x%02x" % frame_type) from exc
        if not isinstance(body, dict):
            raise ProtocolError("frame 0x%02x body must be an object" % frame_type)
        return frame_type, None, body

    if len(buf) < 5:
        raise ProtocolError("stream frame 0x%02x is missing its id" % frame_type)
    (stream_id,) = struct.unpack(">I", buf[1:5])
    return frame_type, stream_id, buf[5:]


def payload_json(payload: bytes) -> dict:
    """Parses the JSON tail of STREAM_OPEN / STREAM_RESET; tolerates an empty tail."""
    if not payload:
        return {}
    try:
        body = json.loads(payload.decode("utf-8"))
    except (ValueError, UnicodeDecodeError):
        return {}
    return body if isinstance(body, dict) else {}

File: client-node/src/protocol.ts
/**
 * Wire protocol v1 — see docs/PROTOCOL.md.
 *
 * This is an independent implementation, not an import from the server: the client is
 * installable on its own and must not drag the server package along. `docs/PROTOCOL.md`
 * is the contract both sides answer to, and `test/protocol.test.ts` checks this encoder
 * against the shared byte vectors in `tests/protocol-vectors.json` — the same vectors the
 * Python client is tested against. Change the protocol there first.
 */

export const PROTOCOL_VERSION = 1;

export const FrameType = {
  Hello: 0x01,
  HelloOk: 0x02,
  Error: 0x03,
  OpenTunnel: 0x10,
  TunnelOk: 0x11,
  CloseTunnel: 0x12,
  StreamOpen: 0x20,
  StreamData: 0x21,
  StreamEof: 0x22,
  StreamAck: 0x23,
  StreamReset: 0x24,
} as const;

export const FRAME_NAMES: Record<number, string> = Object.fromEntries(
  Object.entries(FrameType).map(([name, value]) => [value, name]),
);

const STREAM_FRAME_FLOOR = 0x20;

export class ProtocolError extends Error {}

export function encodeControl(type: number, body: unknown): Buffer {
  const json = Buffer.from(JSON.stringify(body), "utf8");
  const out = Buffer.allocUnsafe(1 + json.length);
  out[0] = type;
  json.copy(out, 1);
  return out;
}

export function encodeStreamJson(type: number, streamId: number, body: unknown): Buffer {
  const json = Buffer.from(JSON.stringify(body), "utf8");
  const out = Buffer.allocUnsafe(5 + json.length);
  out[0] = type;
  out.writeUInt32BE(streamId, 1);
  json.copy(out, 5);
  return out;
}

export function encodeStreamData(streamId: number, payload: Buffer): Buffer {
  const out = Buffer.allocUnsafe(5 + payload.length);
  out[0] = FrameType.StreamData;
  out.writeUInt32BE(streamId, 1);
  payload.copy(out, 5);
  return out;
}

/** For frames whose whole payload is the stream id, e.g. STREAM_EOF. */
export function encodeStreamOnly(type: number, streamId: number): Buffer {
  const out = Buffer.allocUnsafe(5);
  out[0] = type;
  out.writeUInt32BE(streamId, 1);
  return out;
}

export function encodeStreamAck(streamId: number, bytes: number): Buffer {
  const out = Buffer.allocUnsafe(9);
  out[0] = FrameType.StreamAck;
  out.writeUInt32BE(streamId, 1);
  out.writeUInt32BE(bytes, 5);
  return out;
}

export type Frame =
  | { kind: "control"; type: number; body: Record<string, unknown> }
  | { kind: "stream"; type: number; streamId: number; payload: Buffer };

export function decodeFrame(buf: Buffer): Frame {
  if (buf.length < 1) throw new ProtocolError("empty frame");
  const type = buf[0]!;

  if (type < STREAM_FRAME_FLOOR) {
    let body: unknown = {};
    if (buf.length > 1) {
      try {
        body = JSON.parse(buf.subarray(1).toString("utf8"));
      } catch {
        throw new ProtocolError(`frame 0x${type.toString(16)} has a malformed JSON body`);
      }
    }
    if (body === null || typeof body !== "object" || Array.isArray(body)) {
      throw new ProtocolError(`frame 0x${type.toString(16)} body must be a JSON object`);
    }
    return { kind: "control", type, body: body as Record<string, unknown> };
  }

  if (buf.length < 5) {
    throw new ProtocolError(`stream frame 0x${type.toString(16)} is missing its stream id`);
  }
  return { kind: "stream", type, streamId: buf.readUInt32BE(1), payload: buf.subarray(5) };
}

/** Parses the JSON tail of a stream frame that carries one (STREAM_OPEN, STREAM_RESET). */
export function streamFrameJson(payload: Buffer): Record<string, unknown> {
  if (payload.length === 0) return {};
  try {
    const parsed = JSON.parse(payload.toString("utf8"));
    if (parsed === null || typeof parsed !== "object" || Array.isArray(parsed)) return {};
    return parsed as Record<string, unknown>;
  } catch {
    return {};
  }
}
